lib/ansible/modules/packaging/language/pip.py

```diff
@@ -234,6 +260,94 @@
+    and version specifier, providing fallback if the installed setuptools version
+    doesn't support required features.
+    """
+
+    def __init__(self, name_string, version_string=None):
+        self._plain_distribution = False
+        self._distribution_name = name_string
+        if name_string.startswith('file:') or _is_vcs_url(name_string):
+            return
+
+        self._plain_distribution = True
+        if version_string:
+            version_string = version_string.lstrip()
+            separator = '==' if version_string[0].isdigit() else ' '
+            name_string = separator.join((name_string, version_string))
+        self._requirement = Requirement.parse(name_string)
+        # old pkg_resource will replace 'setuptools' with 'distribute' when it already installed
+        if self._requirement.project_name == "distribute":
+            self._distribution_name = "setuptools"
+        else:
+            self._distribution_name = self._requirement.project_name
+
+    @property
+    def has_version_specifier(self):
+        if self._plain_distribution:
+            return bool(self._requirement.specs)
+        return False
+
+    @property
+    def distribution_name(self):
+        return self._distribution_name
+
+    def is_satisfied_by(self, version_to_test):
+        if not self._plain_distribution:
+            return False
+        try:
+            return self._requirement.specifier.contains(version_to_test)
+        except AttributeError:
+            # old setuptools has no specifier, do fallback
+            version_to_test = LooseVersion(version_to_test)
+            return all(
+                op_dict[op](version_to_test, LooseVersion(ver))
+                for op, ver in self._requirement.specs
+            )
+
+    def __str__(self):
+        if self._plain_distribution:
+            return str(self._requirement)
+        return self._distribution_name
+
+
+def _is_vcs_url(name):
+    """Test whether a name is a vcs url or not."""
+    return re.match(r'(svn|git|hg|bzr)\+', name)
+
+
+def _is_valid_distribution_name(name):
+    """Test whether a name is distribution name or version specifier."""
+    return not name.lstrip().startswith(tuple(op_dict.keys()))
+
+
+def _recover_distribution_name(names):
+    """Recover distribution names from user's raw input."""
+    distribution_parts = []
+    for name in names:
+        if _is_valid_distribution_name(name):
+            if len(distribution_parts) > 0:
+                yield ",".join(distribution_parts)
+            distribution_parts = []
+        distribution_parts.append(name)
+    yield ",".join(distribution_parts)
+
+
+def _get_distributions_from_names(module, names):
+    """Parse distribution name to instances of Distribution."""
+    try:
+        return [Distribution(name) for name in _recover_distribution_name(names)]
+    except ValueError as e:
+        module.fail_json(msg=str(e))
```

Comment:
Can you give an example for the error message coming from this? As stated above I would rather remove support for construtcs, which can raise exceptions here.

Reply:
This error comes from the `__init__` of `Distribution`, since the user may input some invalid package name like `djan--go`. The `Requirement.parse(name_string)` will find those errors and raise an error.